Allow volatile access to non-Rust memory, including address 0

LuigiPiucco opened the pull request:
This PR relaxes the `ub_check` in the `read_volatile`/`write_volatile` pointer operations to allow passing null. This is needed to support processors which hard-code peripheral registers on address 0, like the AVR chip ATtiny1626. LLVM understands this as valid and handles it correctly, as tested in my [PR to add a note about it](https://github.com/llvm/llvm-project/pull/139803/commits/6387c82255c56d3035d249eb54110695e76b8030#diff-81bbb96298c32fa901beb82ab3b97add27a410c01d577c1f8c01000ed2055826) (rustc generates the same LLVM IR as expected there when this PR is applied, and consequently the same AVR assembly).

Follow-up and implementation of the discussions in:
- https://internals.rust-lang.org/t/pre-rfc-conditionally-supported-volatile-access-to-address-0/12881/7
- https://github.com/Rahix/avr-device/pull/185;
- [#t-lang > Adding the possibility of volatile access to address 0](https://rust-lang.zulipchat.com/#narrow/channel/213817-t-lang/topic/Adding.20the.20possibility.20of.20volatile.20access.20to.20address.200/with/513303502)
- https://discourse.llvm.org/t/rfc-volatile-access-to-non-dereferenceable-memory-may-be-well-defined/86303

r? @RalfJung

Also fixes https://github.com/rust-lang/unsafe-code-guidelines/issues/29 (about as good as it'll get, null will likely never be a "normal" address in Rust)

LuigiPiucco committed:
fix: don't panic on volatile access to null

According to
https://discourse.llvm.org/t/rfc-volatile-access-to-non-dereferenceable-memory-may-be-well-defined/86303/4,
LLVM allows volatile operations on null and handles it correctly. This
should be allowed in Rust as well, because I/O memory may be hard-coded
to address 0 in some cases, like the AVR chip ATtiny1626.

A test case that ensured a failure when passing null to volatile was
removed, since it's now valid.

Due to the addition of `maybe_is_aligned` to `ub_checks`,
`maybe_is_aligned_and_not_null` was refactored to use it.

docs: revise restrictions on volatile operations

A distinction between usage on Rust memory vs. non-Rust memory was
introduced. Documentation was reworded to explain what that means, and
make explicit that:

- No trapping can occur from volatile operations;
- On Rust memory, all safety rules must be respected;
- On Rust memory, the primary difference from regular access is that
  volatile always involves a memory dereference;
- On Rust memory, the only data affected by an operation is the one
  pointed to in the argument(s) of the function;
- On Rust memory, provenance follows the same rules as non-volatile
  access;
- On non-Rust memory, any address known to not contain Rust memory is
  valid (including 0 and usize::MAX);
- On non-Rust memory, no Rust memory may be affected (it is implicit
  that any other non-Rust memory may be affected, though, even if not
  referenced by the pointer). This should be relevant when, for example,
  reading register A causes a flag to change in register B, or writing
  to A causes B to change in some way. Everything affected mustn't be
  inside an allocation.
- On non-Rust memory, provenance is irrelevant and a pointer with none
  can be used in a valid way.

fix: don't lint null as UB for volatile

Also remove a now-unneeded `allow` line.

fix: additional wording nits

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15119196283/job/42497416225) [(plain)](https://github.com/rust-lang/rust/commit/7ea9ad75460b1362ab83e62bf52b5e5a33c36045/checks/42497416225/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 3.519 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.064 Collecting virtualenv
#12 4.113   Downloading virtualenv-20.31.2-py3-none-any.whl (6.1 MB)
#12 4.346      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.1/6.1 MB 26.2 MB/s eta 0:00:00
#12 4.408 Collecting platformdirs<5,>=3.9.1
#12 4.419   Downloading platformdirs-4.3.8-py3-none-any.whl (18 kB)
#12 4.459 Collecting filelock<4,>=3.12.2
#12 4.466   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.489 Collecting distlib<1,>=0.3.7
#12 4.496   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.507      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 51.0 MB/s eta 0:00:00
#12 4.589 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.782 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.8 virtualenv-20.31.2
#12 4.783 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.9s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      145344 kB
DirectMap2M:     8243200 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-05-12/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
 #[inline]
 #[rustc_allow_const_fn_unstable(const_eval_select)]
-pub(crate) const fn maybe_is_aligned(
-    ptr: *const (),
-    align: usize,
-) -> bool {
+pub(crate) const fn maybe_is_aligned(ptr: *const (), align: usize) -> bool {
     // This is just for safety checks so we can const_eval_select.
     const_eval_select!(
         @capture { ptr: *const (), align: usize } -> bool:
fmt: checked 6021 files
Build completed unsuccessfully in 0:01:16
  local time: Mon May 19 17:27:17 UTC 2025
  network time: Mon, 19 May 2025 17:27:17 GMT
##[error]Process completed with exit code 1.
```

</details>

RalfJung commented:
Cc @rust-lang/opsem

RalfJung commented on library/core/src/ptr/mod.rs:
```diff
@@ -1749,7 +1749,9 @@ pub const unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
-/// operations.
+/// operations. In addition, note that bit-address 0 is legal and can be used
+/// when it's known to hold I/O memory, so using these operations on a null
+/// pointer won't panic.
```
This shouldn't focus on 0. Instead, there should be a dedicated paragraph explaining that
- If a volatile access is performed on regular Rust memory (memory in-bounds of an "allocated object", linking to [here](https://doc.rust-lang.org/nightly/std/ptr/index.html#allocated-object); note that we are in the process of renaming that to "allocation") -- then it is subject to all the same rules wrt aliasing and data races and so.
- However, this method can also be used to access memory that is *not* inside of an allocated object. In that case, it is still UB for the access to trap. However, otherwise this may access arbitrary addresses outside of Rust allocated objects, including 0 and `usize::MAX`.

This requires far-reaching changes in these docs, e.g. the requirement that "src must be valid for reads" only applies in the first of the above cases.

LLVM in its current permits writes to trap, IIUC. We could decide to expose this to our users as well, under the assumption that LLVM won't be able to just take back this guarantee now. It only applies to writes though.

nikic replied:
> LLVM in its current permits writes to trap, IIUC. We could decide to expose this to our users as well, under the assumption that LLVM won't be able to just take back this guarantee now. It only applies to writes though.

From the LLVM side, this is considered something of a hack, not semantics we actually want volatile writes to have. I don't think that Rust should allow this.

workingjubilee replied:
Seems fine! If someone wants to `abort()` then we can just let them do that.

LuigiPiucco replied:
Based on my understanding of your comments, I expanded the scope of the second commit to be a general revision of the documentation on volatile. If anything is incorrect or I exaggerated on the number of changes, I can try to correct/reduce them. If you agree with expanding the scope, I'll also change the title of the PR later. The commit message contains details on the reasoning.

RalfJung commented on library/core/src/ptr/mod.rs:
```diff
@@ -17,6 +17,9 @@
 //!   pointer. The following points are only concerned with non-zero-sized accesses.
+//! * Volatile operations may act on memory outside Rust allocations, such as
+//!   I/O registers at hard-coded addresses. This case is exempt from the rules
+//!   below, but it does have its own rules documented at each volatile operation.
```
Hm, I think I'd prefer to keep these docs here exclusive to actual Rust memory, and say everything volatile-specific in the volatile docs.

LuigiPiucco replied:
I thought leaving the documentation as-was would open the door for ambiguity on whether these rules apply to volatile as well, so I added this point to specify it applies when acting on Rust memory, and otherwise has its own rules. I think we could remove this here, though, and say on the volatile docs that the module rules only apply to volatile in particular cases.

RalfJung replied:
The edits I pushed already do that, by saying that the ptr needs to be valid for reads *or* it can be entirely outside Rust memory.

RalfJung reviewed (commented):
I added a question, and I also pushed a commit where I did an edit pass over the `read_volatile` docs. If you agree with what I did there, could you apply the same edits to `write_volatile`?

RalfJung commented:
@rustbot author

RalfJung commented:
It'd help if you didn't amend and force-push my commit, then I could much easier see what you changed on top of my changes...

RalfJung commented on library/core/src/ptr/mod.rs:
```diff
@@ -1744,54 +1749,65 @@ pub const unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
+///   memory. Here, any address value is possible, including 0 and [`usize::MAX`], so long as the
+///   semantics of such a read are well-defined by the target hardware. The access must not trap. It
+///   can (and usually will) cause side-effects, but those must not affect Rust-allocated memory in
+///   any way. In this use-case, the pointer validity rules in the [module documentation][mod-docs]
+///   may not apply.
```
It seems odd to link to a different place here than where `[valid]` links below, and to use a different term.

```suggestion
///   any way. In this use-case, the pointer does *not* have to be [valid] for reads.
```

LuigiPiucco commented:
Sorry, I'll add new commits instead of amending from now on.

Other projects I worked on seemed to prefer amending, that's why I was doing it like that.

RalfJung commented:
Yeah, every project finds their own way to work around github's deficiencies. :/  I've not yet seen a good way to deal with github's abysmal treatment of force-pushes, though. I often have to pretty much re-review a PR as github provides 0 support for figuring out what actually changes in a series of pushes and force-pushes.

LuigiPiucco commented:
@rustbot ready

If the PR needs further revision, would you prefer to run the `author`/`ready` sequence every time, or to avoid it if we're both replying quickly? I don't mind the pings, but if you do I can adjust.

RalfJung commented:
An explicit "ready for review" signal is always a good idea. :)

RalfJung reviewed (commented):
Good point about the inter-thread synchronization, let's repeat that in the function-level docs. As usual, my `read` comments apply to `write` as well.

RalfJung commented on library/core/src/ptr/mod.rs:
```diff
@@ -112,6 +113,10 @@
 //! fully contiguous (i.e., has no "holes"), there is no guarantee that this
 //! will not change in the future.
 //!
+//! Allocated objects must behave like "normal" memory: in particular, reads must not have
+//! side-effects, and writes must become visible to other threads using the usual synchronization
+//! primitives.
```
Yeah the two PRs conflict anyway, whoever lands second needs to update this.

LuigiPiucco replied:
I already used "allocation" in the places I changed, so I'll adopt it throughout this file in the coming commits.

JakobDegen reviewed (approved):
Just nits

RalfJung commented on library/core/src/ptr/mod.rs:
```diff
@@ -1744,54 +1749,63 @@ pub const unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
-/// incorrect.
+/// Like [`read`], `read_volatile` creates a bitwise copy of `T`, regardless of whether `T` is
+/// [`Copy`]. If `T` is not [`Copy`], using both the returned value and the value at `*src` can
+/// [violate memory safety][read-ownership]. However, modeling volatile memory with non-[`Copy`]
```
Ah good catch, I hadn't even noticed that this paragraph got changed when it got moved around.

This should just keep the exact same wording as before.

LuigiPiucco commented on library/core/src/ub_checks.rs:
```diff
@@ -129,6 +129,25 @@ pub(crate) const fn maybe_is_aligned_and_not_null(
+/// check is anyway not executed in `const`.
+#[inline]
+#[rustc_allow_const_fn_unstable(const_eval_select)]
+pub(crate) const fn maybe_is_aligned(ptr: *const (), align: usize) -> bool {
```
`maybe_is_aligned_and_not_null` is used 12 times across the files `ptr/mod.rs`, `slice/raw.rs` and `intrinsics/mod.rs`. Should I change all of them to your suggestion?

RalfJung replied:
Hm... maybe just implement it in terms of `maybe_is_aligned` then? The ZST thing is subtle enough that it may be worth keeping the function around. But we can deduplicate the `const_eval_select`.

LuigiPiucco commented:
@rustbot ready

I tried to apply all comments and make this PR not conflict with the one changing to allocation. They should be able to apply regardless of order now.

RalfJung reviewed (commented):
On top of these comments, maybe we should also say something about provenance for the "access address outside the machine" case. After all, we probably don't want to permit that when using a pointer that has a well-defined provenance. Such a pointer pretty much has to be created with `with_exposed_provenance`, right?

LuigiPiucco commented on library/core/src/ptr/mod.rs:
Ok. Should I revert the changes related to allocated object -> allocation now, or leave as-is?

RalfJung replied:
Please keep them out of this PR so that we can focus the review on what actually changed here without unnecessary distractions.

LuigiPiucco commented on library/core/src/ptr/mod.rs:
```diff
@@ -1744,54 +1751,66 @@ pub const unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
+///   commonly at fixed addresses reserved by the hardware. These often have special semantics
+///   associated to their manipulation, and cannot be used as general purpose memory. Here, any
+///   address value is possible, including 0 and [`usize::MAX`], so long as the semantics of such a
+///   read are well-defined by the target hardware. The access must not trap. It can (and usually
+///   will) cause side-effects, but those must not affect Rust-allocated memory in in any way. This
```
In the embedded world, it's frequent enough to justify this comment, but since the scope expanded to include volatile as a whole, maybe I should remove it.

LuigiPiucco commented:
About provenance, I'll do some reading to try and propose a comment on its interaction with volatile, but it may be something I'm not fully qualified to do. I'll first push the other requested changes.

LuigiPiucco commented:
I've been thinking about the interaction with provenance, and it seems to me that adding a section as below to each volatile operation should be sufficient. The reasoning for non-AM memory is backed by a comment in the module-level documentation (it's also present in the documentation of `with_exposed_provenance`), which states that "memory which is outside the control of the Rust abstract machine (MMIO registers, for example) is always considered to be accessible with an exposed provenance, so long as this memory is disjoint from memory that will be used by the abstract machine such as the stack, heap, and statics". For AM memory, I assumed that, since volatile access should respect all the rules of regular non-volatile access, provenance would behave the same, thus allowing Exposed and Strict Provenance just like regular access. If this wording is agreed upon, I will add it in a commit.

I do wonder if `without_provenance` from Strict Provenance wouldn't also make sense for MMIO access, though. It seems to me that it would be an alternative way of framing the "from thin-air" provenance, and at first glance, it appears valid.

---
# Provenance

Since provenance describes the validity of accesses a pointer can be used for, this may include information relating to whether volatile accesses are permitted.

For access inside an allocation, the provenance of the pointer is treated the same as regular access, and both Exposed Provenance and Strict Provenance APIs can be used. This is because volatile is more restricted in how the operation will be lowered, so if the pointer's provenance satisfies validity constraints for a regular access, it's unlikely that forcing it to work on actual memory will make it invalid.

For volatile non-Rust accesses, like MMIO, only Exposed Provenance applies; such accesses will inevitably occur at a pointer created "from thin-air", or at one derived from such a pointer, because if it has provenance derived from a Rust allocation, naturally it falls under the use-case above. In other words, the `with_exposed_provenance` call will **not** have a corresponding `expose_provenance` call, instead relying on the fact that memory outside any allocation is always exposed.

LuigiPiucco commented:
> By "MMIO access", do you mean "access to non-Rust memory"?

Yes, exactly. I'll push a simpler edit, then.

LuigiPiucco commented:
Would it be OK to change the Git history like in [this branch](https://github.com/LuigiPiucco/rust/tree/volatile-fixup)? The tip is the same, I just squashed and reordered some commits, and removed the ones that got reverted (to correct the warning above). I can also rebase onto latest master if that would be good.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15377050341/job/43263105465) [(plain)](https://github.com/rust-lang/rust/commit/ceaef25df1e26bc59e18111041cffb1a8b76b20d/checks/43263105465/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/pass-dep/tokio/file-io.rs ... FAILED
tests/pass-dep/libc/mmap.rs ... ok

FAILED TEST: tests/pass-dep/tokio/file-io.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-n8a8SO" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/tests/pass-dep/tokio" "tests/pass-dep/tokio/file-io.rs" "-Zmiri-disable-isolation" "--extern" "cfg_if=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libcfg_if-55a32e410325a882.rlib" "--extern" "cfg_if=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libcfg_if-55a32e410325a882.rmeta" "--extern" "getrandom_01=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-77190f9b44a3c009.rlib" "--extern" "getrandom_01=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-77190f9b44a3c009.rmeta" "--extern" "getrandom_02=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-56888fca95a09bf8.rlib" "--extern" "getrandom_02=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-56888fca95a09bf8.rmeta" "--extern" "getrandom_03=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-a8fc577bb44b6d80.rlib" "--extern" "getrandom_03=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libgetrandom-a8fc577bb44b6d80.rmeta" "--extern" "libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/liblibc-dd98cac8117d550a.rlib" "--extern" "libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/liblibc-dd98cac8117d550a.rmeta" "--extern" "num_cpus=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libnum_cpus-7ab8becb6da559b7.rlib" "--extern" "num_cpus=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libnum_cpus-7ab8becb6da559b7.rmeta" "--extern" "page_size=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libpage_size-c34469ad90c5eb76.rlib" "--extern" "page_size=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libpage_size-c34469ad90c5eb76.rmeta" "--extern" "tempfile=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libtempfile-d40c54a83e8abae9.rlib" "--extern" "tempfile=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libtempfile-d40c54a83e8abae9.rmeta" "--extern" "tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libtokio-539f2d149eae570a.rlib" "--extern" "tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps/libtokio-539f2d149eae570a.rmeta" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/debug/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/tmp/miri_ui/0/miri/x86_64-unknown-linux-gnu/debug/deps" "--edition" "2021"

error: test got exit status: 101, but expected 0
 = note: the compiler panicked

error: no output was expected
---
   4: test_create_and_write::{closure#0}
             at tests/pass-dep/tokio/file-io.rs:24:5
   5: main::{closure#0}
             at tests/pass-dep/tokio/file-io.rs:14:29
   6: tokio::runtime::park::CachedParkThread::block_on::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>::{closure#0}
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/park.rs:284:60
   7: tokio::task::coop::with_budget::<std::task::Poll<()>, {closure@tokio::runtime::park::CachedParkThread::block_on<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>::{closure#0}}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/task/coop/mod.rs:167:5
   8: tokio::task::coop::budget::<std::task::Poll<()>, {closure@tokio::runtime::park::CachedParkThread::block_on<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>::{closure#0}}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/task/coop/mod.rs:133:5
   9: tokio::runtime::park::CachedParkThread::block_on::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/park.rs:284:31
  10: tokio::runtime::context::blocking::BlockingRegionGuard::block_on::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/context/blocking.rs:66:9
  11: tokio::runtime::scheduler::multi_thread::MultiThread::block_on::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>::{closure#0}
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/scheduler/multi_thread/mod.rs:87:13
  12: tokio::runtime::context::runtime::enter_runtime::<{closure@tokio::runtime::scheduler::multi_thread::MultiThread::block_on<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>::{closure#0}}, ()>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/context/runtime.rs:65:16
  13: tokio::runtime::scheduler::multi_thread::MultiThread::block_on::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/scheduler/multi_thread/mod.rs:86:9
  14: tokio::runtime::Runtime::block_on_inner::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/runtime.rs:370:45
  15: tokio::runtime::Runtime::block_on::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/runtime.rs:342:13
  16: main
             at tests/pass-dep/tokio/file-io.rs:15:5
  17: <fn() as std::ops::FnOnce<()>>::call_once - shim(fn())
             at /checkout/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
---
   4: test_create_and_write::{closure#0}
             at tests/pass-dep/tokio/file-io.rs:24:5
   5: main::{closure#0}
             at tests/pass-dep/tokio/file-io.rs:14:29
   6: tokio::runtime::park::CachedParkThread::block_on::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>::{closure#0}
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/park.rs:284:60
   7: tokio::task::coop::with_budget::<std::task::Poll<()>, {closure@tokio::runtime::park::CachedParkThread::block_on<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>::{closure#0}}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/task/coop/mod.rs:167:5
   8: tokio::task::coop::budget::<std::task::Poll<()>, {closure@tokio::runtime::park::CachedParkThread::block_on<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>::{closure#0}}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/task/coop/mod.rs:133:5
   9: tokio::runtime::park::CachedParkThread::block_on::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/park.rs:284:31
  10: tokio::runtime::context::blocking::BlockingRegionGuard::block_on::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/context/blocking.rs:66:9
  11: tokio::runtime::scheduler::multi_thread::MultiThread::block_on::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>::{closure#0}
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/scheduler/multi_thread/mod.rs:87:13
  12: tokio::runtime::context::runtime::enter_runtime::<{closure@tokio::runtime::scheduler::multi_thread::MultiThread::block_on<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>::{closure#0}}, ()>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/context/runtime.rs:65:16
  13: tokio::runtime::scheduler::multi_thread::MultiThread::block_on::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/scheduler/multi_thread/mod.rs:86:9
  14: tokio::runtime::Runtime::block_on_inner::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/runtime.rs:370:45
  15: tokio::runtime::Runtime::block_on::<{async block@tests/pass-dep/tokio/file-io.rs:12:1: 12:15}>
             at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.44.2/src/runtime/runtime.rs:342:13
  16: main
             at tests/pass-dep/tokio/file-io.rs:15:5
  17: <fn() as std::ops::FnOnce<()>>::call_once - shim(fn())
             at /checkout/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
---

Location:
   /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ui_test-0.29.2/src/lib.rs:369

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-f58affc77002becb` (exit status: 1)
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:05:03
  local time: Sun Jun  1 16:59:09 UTC 2025
  network time: Sun, 01 Jun 2025 16:59:09 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

workingjubilee commented on library/core/src/ptr/mod.rs:
```diff
@@ -1744,54 +1749,67 @@ pub const unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
+///
+/// - When a volatile operation is used for memory inside an [allocation], it behaves exactly like
+///   [`read`], except for the additional guarantee that it won't be elided or reordered (see
+///   above). This implies that the operation will actually access memory and not e.g. be lowered to
+///   a register access or stack pop. Other than that, all the usual rules for memory accesses apply
```
hmm, if the memory is literally on the stack though, wouldn't popping it off the stack be a valid implementation? I think something like "won't be lowered to reusing data from a previous read, such as a register that performed a previous load on that memory" might be clearer, since it describes the actual contract?

LuigiPiucco replied:
I think it wouldn't be valid because a stack pop implies changing the stack pointer as well. Because we're talking about memory on the stack, that naturally falls under the Rust AM case. So if a `read_volatile` was allowed to lower into a pop, we'd have an extra side-effect caused by the read, which is disallowed for volatile on Rust memory.

In addition, I locally set up some preliminary testing for x86_64 and AVR, and it doesn't seem like `read_volatile` generates stack pops for either architecture, even for local variables on the stack. This is not conclusive, of course, but because of the above, it probably shouldn't anyway.

RalfJung replied:
The Rust AM has no concept of a Stack Pointer, so a priori there's no reason why the compiler couldn't use such a lowering.

But if that's undesirable, we should get LLVM to document that they don't do that, and then document the same ourselves.

LuigiPiucco replied:
Lowering into a pop sounds strange to me, intuitively I think of volatile operations mapping to an explicit load/store of some form, with the memory address as a specific part of it in some way. But that may be biased toward the embedded use-case of peripheral I/O. I won't dwell on it, and change the line as requested.

workingjubilee replied:
@LuigiPiucco Based on what you said, I'm slightly worried that you imagine that the following is not allowed to move the stack pointer, because that would be a "side effect":
```rust
let x: i32 = 0;
```

...and my understanding is it absolutely is?

The C contract of `volatile` doesn't say anything about stack operations either. I agree that the compiler will usually execute the enhanced contract you describe, but that seems mostly incidental. If a compiler decided to arrange locals on the stack so it could choose `pop`s more often for the final read from the stack, that seems fine?

workingjubilee replied:
That nattering aside, revised text looks good.

LuigiPiucco replied:
Don't worry, I don't think that. I just got my wires crossed yesterday, and it seemed like the pop would be invalid when using volatile, only.

Just to mention it, when I rebase I'll squash the commit with your suggestion (along with the many fixups) into a single documentation commit.

LuigiPiucco commented on tests/ui/precondition-checks/read_volatile.rs:
```diff
@@ -1,7 +1,7 @@
-//@ revisions: null misaligned
+//@ revisions: misaligned
 
 #![allow(invalid_null_arguments)]
```
I just noticed the line above. I assume it's unused now, since there is no longer a null argument, and if so, I'll remove it. But more importantly, does this mean there is still a check enabled by default that prevents the null usage? I think it may be related to the source snippet below.

https://github.com/rust-lang/rust/blob/2398bd60ef526e686a38a299cc2fa991d8b3c33e/compiler/rustc_lint/src/ptr_nulls.rs#L160-L175

Should I remove `read_volatile`/`write_volatile`?

RalfJung commented on library/core/src/ptr/mod.rs:
```diff
@@ -2021,54 +2026,68 @@ pub const unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
+/// semantics will almost always end up pretty similar to [C11's definition of volatile][c11].
+///
+/// Volatile operations are intended to act on I/O memory. As such, they are considered externally
+/// observable events (just like syscalls), and are guaranteed to not be elided or reordered by the
```
```suggestion
/// observable events (just like syscalls, but less opaque), and are guaranteed to not be elided or reordered by the
```
This seems good to mention to avoid people thinking they can treat this as a compiler barrier

RalfJung commented on library/core/src/ptr/mod.rs:
```diff
@@ -2021,54 +2026,68 @@ pub const unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
-/// and may be ignored.
+/// Performs a volatile read of the value from `src` without moving it.
+///
+/// Rust does not currently have a rigorously and formally defined memory model, so the precise
+/// semantics of what "volatile" means here is subject to change over time. That being said, the
+/// semantics will almost always end up pretty similar to [C11's definition of volatile][c11].
```
I wonder if we still need this paragraph. The next paragraph says basically as much as C says, after all.

@rust-lang/opsem what do you think?

RalfJung replied:
If nobody else has an opinion, I'd say please remove this paragraph.

LuigiPiucco commented on library/core/src/ptr/mod.rs:
```diff
@@ -2021,54 +2026,68 @@ pub const unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
+///   access is still not considered [atomic], and as such it cannot be used for inter-thread
+///   synchronization.
+///
+/// Note that volatile memory operations on zero-sized types (e.g., if a zero-sized type is passed
+/// to `read_volatile`) are noops and may be ignored.
```
I think we can shorten and fully remove the duplicate information about zero-sized types by rewording as (then moving into the paragraph before):
```rust
/// Note that volatile memory operations where T is a zero-sized type are noops and may be ignored.
```
But I'll wait before changing this one, and push whichever edit you think is best.

RalfJung commented:
@scottmcm FWIW I think volatile can be justified in a quite principled way. There's a lot of baggage around volatile and it is misunderstood by many, but at its core is a sensible semantics that I hope we can carve out over time.

Happy to discuss this more at some point in a suitable venue :)

RalfJung commented:
@LuigiPiucco that wording you suggested sounds good. :)

RalfJung commented on tests/ui/lint/invalid_null_args.rs:
Please keep the volatile calls, and explicitly have comments have saying we do *not* want the lint to fire there.

LuigiPiucco replied:
Please check if I did it correctly.

RalfJung replied:
Lint test change looks good :)

RalfJung commented:
@rustbot author

RalfJung commented:
@JakobDegen this changed a bunch since your last review so it probably needs another look.

LuigiPiucco commented:
@RalfJung, I have rebased the branch on master and squashed the history. The only "new" change is removing the c11 link, which became orphaned and I had forgotten to remove it. I say this here because the relevant changes were squashed into the revision commit, to avoid surprise. I'll also re-request Jakob's review through GitHub, since the PR changed a lot.

RalfJung commented on library/core/src/ptr/mod.rs:
```diff
@@ -2021,54 +2026,61 @@ pub const unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
+///   Here, any address value is possible, including 0 and [`usize::MAX`], so long as the semantics
+///   of such a read are well-defined by the target hardware. The provenance of the pointer is
+///   irrelevant, and it can be created with [`without_provenance`]. The access must not trap. It
+///   can cause side-effects, but those must not affect Rust-allocated memory in in any way. This
```
```suggestion
///   can cause side-effects, but those must not affect Rust-allocated memory in any way. This
```

RalfJung reviewed (commented):
I'll take over this one since we haven't heard from @JakobDegen in a bit.
r? @RalfJung 
I just have a few wording nits. :)
@rustbot author

RalfJung commented:
This looks great, thanks! Please [squash the commits](https://rustc-dev-guide.rust-lang.org/git.html#squash-your-commits), then we can land this. Please use the `--keep-base` flag when squashing so that the force-push diff is easier to review; do *not* rebase on the latest master branch. Then write `@rustbot ready` after you force-pushed the squashed PR.

@rustbot author

RalfJung commented:
Looks good, thanks!
@bors r+ rollup

bors commented:
:pushpin: Commit 8a8717e971dbdc6155506a4332e9ce8ef9151caa has been approved by `RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung 8a8717e971dbdc6155506a4332e9ce8ef9151caa -->
<!-- homu: {"type":"Approved","sha":"8a8717e971dbdc6155506a4332e9ce8ef9151caa","approver":"RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
